File: src/jax/java/jaxdemo/view/PersonRecordView.java
package jax.java.jaxdemo.view;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import jax.java.jaxdemo.eventhandler.EventHandler;
import jax.java.jaxdemo.events.PersonRecordEvent;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.joda.time.DateTime;

public class PersonRecordView implements EventHandler<PersonRecordEvent> {
	private Map<String, PersonRecord> viewStore = new HashMap<String, PersonRecord>(1000000);
	private Map<String, PersonRecord> ssnIndex = new HashMap<String, PersonRecord>(1000000);
	private Map<DateTime, List<PersonRecord>> dateIndex = new HashMap<DateTime, List<PersonRecord>>(100000);
	private Map<String, String> idMap = new HashMap<String, String>();
	private Logger log = LogManager.getLogger(this);

	public Map<String, PersonRecord> getViewStore() {
		return viewStore;
	}

	public Map<String, PersonRecord> getSsnIndex() {
		return ssnIndex;
	}

	public Map<DateTime, List<PersonRecord>> getDateIndex() {
		return dateIndex;
	}

	private int currentId = 0;
	
	private AtomicInteger eventCount = new AtomicInteger(0);

	@Override
	public void applyEvent(PersonRecordEvent personRecordEvent) {

		String ssn = personRecordEvent.getSsn();
		DateTime dob = personRecordEvent.getDob();
		String id = getId(ssn);
		PersonRecord personRecord = new PersonRecord(id, personRecordEvent.getFirstName(), personRecordEvent.getMiddleInitial(),
				personRecordEvent.getLastName(), ssn, dob);

		PersonRecord previousPersonRecord = viewStore.get(id);
		viewStore.put(id, personRecord);
		ssnIndex.put(ssn, personRecord);
		removePreviousPersonFromDateIndex(previousPersonRecord);
		addToDateIndex(dob, personRecord);
		
		int currentEventCount = eventCount.incrementAndGet();
		if (currentEventCount % 250000 ==0){
			log.info("Processed event ("+currentEventCount+"): "+ personRecordEvent);
		}

	}

	private void removePreviousPersonFromDateIndex(PersonRecord previousPersonRecord) {
		if (previousPersonRecord != null && dateIndex.containsKey(previousPersonRecord.getDob())) {
			dateIndex.get(previousPersonRecord.getDob()).remove(previousPersonRecord);
		}
	}

	private void addToDateIndex(DateTime dob, PersonRecord personRecord) {
		List<PersonRecord> personRecordByDate = dateIndex.get(dob);
		if (personRecordByDate == null) {
			personRecordByDate = new ArrayList<PersonRecord>();
			dateIndex.put(dob, personRecordByDate);
		}
		personRecordByDate.add(personRecord);
	}

	synchronized private String getId(String ssn) {

		String id = idMap.get(ssn);
		if (id == null) {
			currentId++;
			id = Integer.toString(currentId);
			idMap.put(ssn, id);
		}

		return id;

	}

}

File: src/jax/scala/jaxdemo/view/PersonRecordView.scala
package jax.scala.jaxdemo.view

import jax.scala.jaxdemo.eventhandler.EventHandler
import jax.scala.jaxdemo.events.PersonRecordEvent
import scala.collection.mutable.HashMap
import org.joda.time.DateTime
import java.util.concurrent.atomic.AtomicInteger
import scala.collection.mutable.ListBuffer
import org.apache.logging.log4j.LogManager
case class PersonRecordView extends EventHandler[PersonRecordEvent] {

  val log = LogManager.getLogger(this)
  val viewStore = new HashMap[String, PersonRecord]
  val ssnIndex = new HashMap[String, PersonRecord]
  val dateIndex = new HashMap[DateTime, ListBuffer[PersonRecord]]
  val idMap = new HashMap[String, String]
  val currentId: AtomicInteger = new AtomicInteger(0)
  val eventCount: AtomicInteger = new AtomicInteger(0)

  override def applyEvent(personRecordEvent: PersonRecordEvent) {
    val ssn = personRecordEvent.ssn
    val id = getId(ssn)
    val dob = personRecordEvent.dob

    val personRecord = new PersonRecord(id, personRecordEvent.firstName, personRecordEvent.middleInitial, personRecordEvent.lastName, ssn, dob);
    val previousPersonRecord = viewStore.get(id);
    removePreviousPersonFromDateIndex(previousPersonRecord)
    viewStore.put(id, personRecord);
    ssnIndex.put(ssn, personRecord);
    addToDateIndex(dob, personRecord)
    logEvent(personRecordEvent)

  }

  private def getId(ssn: String) = {
    idMap.getOrElseUpdate(ssn, String.valueOf(currentId.incrementAndGet()))
  }

  private def removePreviousPersonFromDateIndex(previousPersonRecordOption: Option[PersonRecord]) {
    val previousPersonRecord = previousPersonRecordOption.getOrElse[PersonRecord](PersonRecord())
    val personRecordByDate = dateIndex.getOrElse(previousPersonRecord.dob, new ListBuffer[PersonRecord])
    personRecordByDate -= previousPersonRecord
  }

  private def addToDateIndex(dob: DateTime, personRecord: PersonRecord) {
    val personRecordByDate = dateIndex.getOrElseUpdate(dob, new ListBuffer[PersonRecord])
    personRecordByDate += personRecord
  }

  private def logEvent(personRecordEvent: PersonRecordEvent) {
    val currentEventCount = eventCount.incrementAndGet()
    if (currentEventCount % 250000 == 0) {
      log.info(s"Processed event $currentEventCount: $personRecordEvent")
    }
  }

}